ergotree-ir/src/chain/header.rs

```diff
@@ -42,23 +49,219 @@ pub struct Header {
+#[cfg_attr(feature = "json", derive(serde::Serialize, serde::Deserialize))]
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct AutolykosSolution {
     /// Public key of miner. Part of Autolykos solution.
-    #[cfg_attr(feature = "json", serde(skip_serializing, skip_deserializing))]
+    #[cfg_attr(feature = "json", serde(rename = "pk"))]
     pub miner_pk: Box<dlog_group::EcPoint>,
     /// One-time public key. Prevents revealing of miners secret.
-    #[cfg_attr(feature = "json", serde(skip_serializing, skip_deserializing))]
-    pub pow_onetime_pk: Box<dlog_group::EcPoint>,
+    #[cfg_attr(feature = "json", serde(default, rename = "w"))]
+    pub pow_onetime_pk: Option<Box<dlog_group::EcPoint>>,
     /// nonce
-    #[cfg_attr(feature = "json", serde(skip_serializing, skip_deserializing))]
+    #[cfg_attr(
+        feature = "json",
+        serde(
+            rename = "n",
+            serialize_with = "crate::chain::json::autolykos_solution::as_base16_string",
+            deserialize_with = "crate::chain::json::autolykos_solution::from_base16_string"
+        )
+    )]
     pub nonce: Vec<u8>,
     /// Distance between pseudo-random number, corresponding to nonce `nonce` and a secret,
-    /// corresponding to `miner_pk`. The lower `pow_distance` is, the harder it was to find this solution.
-    #[cfg_attr(feature = "json", serde(skip_serializing, skip_deserializing))]
-    pub pow_distance: BigInt,
-    /// Miner votes for changing system parameters.
-    /// 3 bytes in accordance to Scala implementation, but will use `Vec` until further improvements
-    #[cfg_attr(feature = "json", serde(rename = "votes"))]
-    pub votes: Votes,
+    /// corresponding to `miner_pk`. The lower `pow_distance` is, the harder it was to find this
+    /// solution.
+    ///
+    /// Note: we serialize/deserialize through custom functions since `BigInt`s serde implementation
+    /// encodes the sign and absolute-value of the value separately, which is incompatible with the
+    /// JSON representation used by Ergo. ASSUMPTION: we assume that `pow_distance` encoded as a
+    /// `u64`.
+    #[cfg_attr(
+        feature = "json",
+        serde(
+            default,
+            rename = "d",
+            serialize_with = "crate::chain::json::autolykos_solution::bigint_as_str",
+            deserialize_with = "crate::chain::json::autolykos_solution::bigint_from_serde_json_number"
+        )
+    )]
+    pub pow_distance: Option<BigInt>,
+}
+
+impl AutolykosSolution {
+    fn serialize_bytes<W: WriteSigmaVlqExt>(
+        &self,
+        version: u8,
+        w: &mut W,
+    ) -> Result<(), ScorexSerializationError> {
+        if version == 1 {
+            self.miner_pk.scorex_serialize(w)?;
+            #[allow(clippy::unwrap_used)]
+            self.pow_onetime_pk.as_ref().unwrap().scorex_serialize(w)?;
+            w.write_all(&self.nonce)?;
+
+            // pow_distance must be == Some(_) for autolykos v1.
+            #[allow(clippy::unwrap_used)]
```

Comment:
Can remove these 2 unwraps by adding a `Misc` variant to `ScorexSerializationError`. What do you think?

Reply:
Good call! Let's be extra safe here.